rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -583,15 +645,43 @@ object ExecutionBatch extends Loggable {
                                    missing ++ unexpected ++ expected
                                 }
     } yield {
-      RuleNodeStatusReport(
-          mergeInfo.nodeId
-        , ruleId
-        , serial
-        , mergeInfo.run
-        , mergeInfo.configId
-        , DirectiveStatusReport.merge(directiveStatusReports)
+      (
+          SerialedRuleId(ruleId, serial)
+        , RuleNodeStatusReport(
+              mergeInfo.nodeId
+            , ruleId
+            , serial //can not be empty because of groupBy
+            , mergeInfo.run
+            , mergeInfo.configId
+            , DirectiveStatusReport.merge(directiveStatusReports)
+            , mergeInfo.expirationTime
+          )
       )
-    }).toSeq
+    }).toMap
+
+    //now, for all current expected reports, choose between the computed value and the default one
+
+    // note: isn't there something specific to do for unexpected reports ? Keep them all ?
+
+    val nil = Seq[RuleNodeStatusReport]()
+    val (computed, newStatus) = ((nil, nil)/: currentExpectedDirectives) { case ( (c,n), (k, expectedReport)) =>
+      complianceForRun.get(k) match {
+        case None => //the whole rule is new!
+          //here, the reports are ACTUALLY pending, not missing.
+          val x = buildRuleNodeStatusReport(mergeInfo, Map(k -> expectedReport), PendingReportType)
```

Comment:
The expiration status is at the report level, not individual component, because expiration is linked to a not sending report or not. 

That last part of the algorithm is managing things (rule for now, directive in the future) whose serial changed since the last run. If serial changed, it means config was updated, and since we dealt with all the macro status (unexpect version, etc) before, the status is pending and not missing or success. 

k is just the name of the map _k_eys (don't care much of what they are, ruleAndserialId certainly)